File: src/rag-module/bm25.ts
import { tokenize } from "./utils";

type DocField = {
  qid: string;
  text: string;
};

type BM25Row = {
  qid: string;
  score: number;
};

export class BM25Index {
  private docs = new Map<string, string[]>();
  private docLen = new Map<string, number>();
  private inverted = new Map<string, Array<{ qid: string; tf: number }>>();
  private avgDocLen = 0;
  private totalDocs = 0;
  private k1 = 1.2;
  private b = 0.75;

  addDocuments(rows: DocField[]): void {
    for (const row of rows) {
      const tokens = tokenize(row.text);
      this.docs.set(row.qid, tokens);
    }
    this.rebuildInverted();
  }

  removeDocuments(qids: string[]): void {
    for (const qid of qids) {
      this.docs.delete(qid);
    }
    this.rebuildInverted();
  }

  private rebuildInverted(): void {
    this.inverted.clear();
    this.docLen.clear();
    let totalLen = 0;

    for (const [qid, tokens] of this.docs) {
      totalLen += tokens.length;
      this.docLen.set(qid, tokens.length);

      const tf = new Map<string, number>();
      for (const token of tokens) tf.set(token, (tf.get(token) ?? 0) + 1);

      for (const [term, count] of tf.entries()) {
        const posting = this.inverted.get(term) ?? [];
        posting.push({ qid, tf: count });
        this.inverted.set(term, posting);
      }
    }

    this.totalDocs = this.docs.size;
    this.avgDocLen = this.totalDocs ? totalLen / this.totalDocs : 0;
  }

  search(query: string, topK: number): BM25Row[] {
    const qTokens = tokenize(query);
    if (!qTokens.length || this.totalDocs === 0) return [];

    const queryTerms = new Set(qTokens);
    const scores = new Map<string, number>();

    for (const term of queryTerms.values()) {
      const posting = this.inverted.get(term);
      if (!posting || posting.length === 0) continue;

      const df = posting.length;
      const idf = Math.log(1 + (this.totalDocs - df + 0.5) / (df + 0.5));

      for (const { qid, tf } of posting) {
        const dl = Math.max(this.docLen.get(qid) ?? 0, 1);
        const numerator = tf * (this.k1 + 1);
        const denominator = tf + this.k1 * (1 - this.b + this.b * (dl / Math.max(this.avgDocLen, 1)));
        const add = idf * (numerator / denominator);
        scores.set(qid, (scores.get(qid) ?? 0) + add);
      }
    }

    return Array.from(scores.entries())
      .map(([qid, score]) => ({ qid, score }))
      .sort((a, b) => b.score - a.score)
      .slice(0, topK);
  }
}

File: rag_python/bm25.py
from __future__ import annotations

import math
from collections import defaultdict

from .utils import tokenize


class BM25Index:
    def __init__(self, k1: float = 1.2, b: float = 0.75):
        self.k1 = k1
        self.b = b
        self.docs: dict[str, list[str]] = {}
        self.doc_len: dict[str, int] = {}
        self.inverted: dict[str, list[tuple[str, int]]] = defaultdict(list)
        self.total_docs = 0
        self.avg_doc_len = 0.0

    def add_documents(self, rows: list[dict[str, str]]) -> None:
        for row in rows:
            self.docs[row["qid"]] = tokenize(row["text"])
        self._rebuild()

    def remove_documents(self, qids: list[str]) -> None:
        for qid in qids:
            self.docs.pop(qid, None)
        self._rebuild()

    def _rebuild(self) -> None:
        self.doc_len.clear()
        self.inverted = defaultdict(list)
        total_len = 0

        for qid, tokens in self.docs.items():
            total_len += len(tokens)
            self.doc_len[qid] = len(tokens)
            tf: dict[str, int] = defaultdict(int)
            for t in tokens:
                tf[t] += 1
            for term, count in tf.items():
                self.inverted[term].append((qid, count))

        self.total_docs = len(self.docs)
        self.avg_doc_len = total_len / self.total_docs if self.total_docs else 0.0

    def search(self, query: str, top_k: int) -> list[dict[str, float]]:
        q_terms = set(tokenize(query))
        if not q_terms or self.total_docs == 0:
            return []

        scores: dict[str, float] = defaultdict(float)
        avgdl = max(self.avg_doc_len, 1.0)

        for term in q_terms:
            posting = self.inverted.get(term)
            if not posting:
                continue
            df = len(posting)
            idf = math.log(1 + (self.total_docs - df + 0.5) / (df + 0.5))
            for qid, tf in posting:
                dl = max(self.doc_len.get(qid, 0), 1)
                num = tf * (self.k1 + 1)
                den = tf + self.k1 * (1 - self.b + self.b * (dl / avgdl))
                scores[qid] += idf * (num / den)

        out = [{"qid": qid, "score": score} for qid, score in scores.items() if score > 0]
        out.sort(key=lambda x: x["score"], reverse=True)
        return out[:top_k]
